### DISCUSSION TITLE
UntrustedHost when deploying to Vercel, even if I manually set AUTH_TRUST_HOST=true

### DISCUSSION THREAD

**Discussion by hopperelec** [nextauthjs/next-auth#11725] [Help]
To my understanding, you're not supposed to get UntrustedHost if `trustHost` is truthy, and it's supposed to be automatically truthy when detected as running on Vercel. However, this doesn't seem to be working for me. I thought that maybe it just wasn't detecting Vercel correctly, so I tried manually setting `AUTH_TRUST_HOST=true` but that didn't work either. What could be causing this?

**Comment by hopperelec** [comment]
Still having this issue

**Reply by leeoocca** [reply]
Same here. You found any solutions?

**Reply by hopperelec** [reply]
Haven't looked at this in a while, so maybe there's been some changes since. Otherwise, no, I have not found a solution unfortunately,

**Reply by leeoocca** [reply]
I'm looking into migrating to better-auth. I don't understand how something like this can be around. Seems pretty blocking. By any chance you were also using SvelteKit?

**Reply by hopperelec** [reply]
Yes, I was! It was for [this project](https://github.com/[hopperelec/competitive-replying) specifically, although I never got round to finishing it

**Comment by MILLERMARRU** [comment]
This is a common Auth.js v5 (next-auth@beta) issue on Vercel. There are three separate root causes — here is how to diagnose which one you're hitting.

## Root cause 1 — `AUTH_URL` is set and wrong (most common)

If you have `AUTH_URL` defined in Vercel's environment variables with an incorrect value (e.g., `http://localhost:3000` left over from local dev), it **overrides** the auto-detection and causes `UntrustedHost`.

**Fix:** Go to Vercel → Project → Settings → Environment Variables and either **delete `AUTH_URL`** entirely, or set it to your exact deployment URL:

```
AUTH_URL=https://your-app.vercel.app
```

## Root cause 2 — `AUTH_TRUST_HOST` not propagated to the right environment

Vercel has three environments: Production, Preview, and Development. Make sure `AUTH_TRUST_HOST=true` is set for **all three**, not just Production.

```
AUTH_TRUST_HOST=true   ← set for Production + Preview + Development
```

## Root cause 3 — Explicit config override

As a fallback, set `trustHost` directly in your auth config so it's never dependent on environment detection:

```ts
// auth.ts
import NextAuth from 'next-auth';
import GitHub from 'next-auth/providers/github';

export const { handlers, auth, signIn, signOut } = NextAuth({
  trustHost: true,   // ← explicit, not relying on env var detection
  providers: [GitHub],
});
```

## Why auto-detection sometimes fails

Auth.js v5 auto-enables `trustHost` when the `VERCEL` env var is present (Vercel sets this automatically on all deployments). The failure usually happens when:
- You have `AUTH_URL` set to a wrong value that takes precedence
- You're on a custom domain that wasn't present when the session was created
- A proxy (Vercel's Edge Network, a custom domain, or a reverse proxy) changes the `host` header

The safest setup for Vercel is:

```bash
# Vercel env vars (Production + Preview + Development)
AUTH_SECRET=your-secret-here
AUTH_TRUST_HOST=true
# Do NOT set AUTH_URL unless you have a specific reason to
```
